lib/class-wp-theme-json-gutenberg.php

```diff
@@ -3869,4 +3869,41 @@ function ( $carry, $item ) {
 		$theme_json->theme_json['styles'] = self::convert_variables_to_value( $styles, $vars );
 		return $theme_json;
 	}
+
+	/**
+	 * Generates a selector for a block style variation.
+	 *
+	 * @param string $variation_name Name of the block style variation.
+	 * @param string $block_selector CSS selector for the block.
+	 *
+	 * @return string Block selector with block style variation selector added to it.
+	 */
+	public static function get_block_style_variation_selector( $variation_name, $block_selector ) {
```

Comment:
Does this need to be a `public` function? It looks like most of the similar methods (`append_to_selector`, `prepend_to_selector`, etc) are `protected`.

Reply:
I figured this one warranted some tests given the regex in it and I wanted the ability to generate the variation selector in a hook if we end up going down the per-block instance variations styles path.

Do you think making it generic e.g. `inject_class` or something more appropriate, would be better? Then do the extra leg work to make the variation selectors where they are needed.

Reply:
> Do you think making it generic e.g. inject_class or something more appropriate, would be better? 

If making the method public, then yes, I think it'd be better to try to make it a bit more generic if we can. We're already a bit boxed in with the complexity of the Theme JSON class, but at least if the public API doesn't grow too much, there's the possibility of potentially simplifying things further down the track.

If going for `protected`, it'd still be great to include tests to cover the regex, but I'd probably lean toward optimising for a reduced public API surface area over the ease of testing. Or to put it another way, if we can do effectively the same tests, but moving "up" a couple of levels so we're testing the final output of the Theme JSON class rather than this particular method?

Reply:
Oh, one other thing in case it helps: if it's only set to `public` so that we can test the method, then we might be able to have it be `protected` and then use `ReflectionClass` and set the method to be accessible for your tests. I haven't tried this before, but found the following while looking through tests in `wordpress-develop`: https://github.com/WordPress/wordpress-develop/blob/2a6b527f73260925729beb7860a4a2a01513f527/tests/phpunit/tests/admin/wpSiteHealth.php#L44-L46

And: https://github.com/WordPress/wordpress-develop/blob/4ee594761f16e7aa9314fbf3088c3ad0c9ea3681/tests/phpunit/tests/hooks/removeAllFilters.php#L57-L63

Reply:
You make some good points @andrewserong 👍 

I get the suspicion that we will need to generate a block style variation selector outside of theme.json so the public option might be required. 🤔

I'm a bit short on good names for a more generic function. Do you have any ideas? 

- `add_class_to_ancestor`
- `add_class_to_selector_ancestors`
- `inject_class_in_selector`

In terms of related functions, `scope_selector` is public so I guess that's a precedent.

Reply:
Oh, naming it tough! 😄

I don't have any real preferences, but just throwing a few more into the ring:

* `add_class_to_outer_selector`
* `add_class_to_block_selector` (maybe too specific)
* `add_class_to_wrapper_selector`

Overall, I think I like the `add_class` prefix 👍

ChatGPT gave some similarly repetitive suggestions 🤔

1. `addWrapperClassToSelector`
2. `addClassToWrapperSelector`
3. `appendToWrapperSelector`
4. `updateWrapperSelectorClass`
5. `setWrapperClassInSelector`
6. `modifyWrapperSelectorWithClass`
7. `applyClassToWrapperSelector`
8. `prependClassToWrapperSelector`
9. `insertClassInWrapperSelector`
10. `setAncestorClassInSelector`
11. `addAncestorClassToSelector`
12. `updateAncestorSelectorClass`
13. `applyClassToAncestorSelector`
14. `prependClassToAncestorSelector`
15. `insertClassInAncestorSelector`

Reply:
I think the catch with `add_class_to_block_selector` is that the block's selector could be a compound and comma separated selector. It's why I tried to target the fact it is the ancestor that we're really trying to add the class to.

Now a few more minutes have passed. I'm going to sidestep the public API side of it. The new plan is:

1. Switch `get_block_style_variation_selector` to `protected`
2. Update the tests to call it via reflection class
3. When/if a hook does need that functionality, `get_block_style_variation_selector` can be refactored to have a generic public function do the heavy lifting which just gets used internally in `get_block_style_variation_selector`.

I've done the first to parts of this in https://github.com/WordPress/gutenberg/pull/58051/commits/1032daf873b826eb77529158c49143f30d938007. 

There's still the JS side of it which I'll get to shortly but I think the PHP side should be close now. Thanks for the help.